{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/KULL-Centre/_2023_Tesei_IDRome/blob/main/IDRLab.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PEwBcDpng7K6"
   },
   "source": [
    "## **Preliminary information:**\n",
    "\n",
    "This Colab notebook enables running molecular dynamics (MD) simulations of intrinsically disordered proteins (IDPs) and protein regions (IDRs) and to study their conformational ensembles.\n",
    "\n",
    "MD simulations employ the coarse-grained force field CALVADOS2, where each residue is mapped onto a single bead and modeled with a stickiness parameter and electrostatics.\n",
    "\n",
    "Simulations only require that the user provides the sequence of an IDP, set environmental conditions (temperature & ionic strength), and the charge states of His residues and terminal amino and carboxyl groups.\n",
    "\n",
    "<b><font color='#FA003F'>Please note:</font></b> Cells for preliminary operations should be executed one by one to prevent crashes. This notebook uses condacolab, whose installation will cause a kernel restart. Because of this, a crash will happen during preliminary operations if you execute all cells at once.\n",
    "\n",
    "<b><font color='#FA003F'>How to cite IDRLab:</font></b> If you use IDRLab, we ask you to cite Tesei, Trolle et al. [1]. In addition, you may consider citing Tesei & Lindorff-Larsen [2] for CALVADOS 2 and Eastman et al. [3] for OpenMM.\n",
    "1. G. Tesei, A. I. Trolle, N. Jonsson, J. Betz, F. Pesce, K. E. Johansson, K. Lindorff-Larsen __Conformational ensembles of the human intrinsically disordered proteome: Bridging chain compaction with function and sequence conservation__ _bioRxiv_ 2023 2023.05.08.539815 DOI: https://doi.org/10.1101/2023.05.08.539815\n",
    "\n",
    "2. G. Tesei and K. Lindorff-Larsen __Improved predictions of phase behaviour of intrinsically disordered proteins by tuning the interaction range [version 2; peer review: 2 approved]__ _Open Research Europe_ 2023 2(94) DOI: https://doi.org/10.12688/openreseurope.14967.2\n",
    "3. P. Eastman, J. Swails, J. D. Chodera et al. __OpenMM 7: Rapid development of high performance algorithms for molecular dynamics__ _PLoS Computational Biology_ 2017 13(7):e1005659 DOI: https://doi.org/10.1371/journal.pcbi.1005659\n",
    "---\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "0F7WtrIChg-F"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#E3B505'>1 - Input protein sequence</font></b>\n",
    "import numpy as np\n",
    "import os\n",
    "import shutil\n",
    "import ipywidgets as widgets\n",
    "\n",
    "try:\n",
    "    os.rmdir('sample_data')\n",
    "except:\n",
    "    pass\n",
    "#@markdown Name the IDR that you want to simulate:\n",
    "NAME = \"Q99457_1_97\" #@param {type:\"string\"}\n",
    "\n",
    "#@markdown Insert here the sequence of the IDR that you want to simulate:\n",
    "SEQUENCE = \"MAEADFKMVSEPVAHGVAEEEMASSTSDSGEESDSSSSSSSTSDSSSSSSTSGSSSGSGSSSSSSGSTSSRSRLYRKKRVPEPSRRARRAPLGTNFV\" #@param {type:\"string\"}\n",
    "if \" \" in SEQUENCE:\n",
    "    SEQUENCE = ''.join(SEQUENCE.split())\n",
    "    print('Blank character(s) found in the provided sequence. Sequence has been corrected, but check for integrity:')\n",
    "    print(SEQUENCE)\n",
    "    print('\\n')\n",
    "#@markdown Simulation settings:\n",
    "Temperature = 310 #@param {type:\"number\"}\n",
    "Ionic_strength = 0.150 #@param {type:\"number\"}\n",
    "#@markdown <i>*Units: Temperature [K], Ionic_strength [M]<i>\n",
    "\n",
    "#@markdown Define charge states:\n",
    "charged_N_terminal_amine = True #@param {type:\"boolean\"}\n",
    "charged_C_terminal_carboxyl = False #@param {type:\"boolean\"}\n",
    "charged_histidine = False #@param {type:\"boolean\"}\n",
    "Nc = 1 if charged_N_terminal_amine == True else 0\n",
    "Cc = 1 if charged_C_terminal_carboxyl == True else 0\n",
    "\n",
    "if charged_histidine == True:\n",
    "    print('Define pH and pKa to set the charge of Histidines according to the Henderson-Hasselbalch equation.')\n",
    "    pH = input('Enter pH value: ')\n",
    "    pH = float(pH)\n",
    "    pKa = input('Enter pKa value: ')\n",
    "    pKa = float(pKa)\n",
    "    Hc = 1/(1+10**(pH-pKa))\n",
    "if charged_histidine == False:\n",
    "    Hc = 0\n",
    "\n",
    "np.savetxt('env_settings.txt', np.array([Temperature, Ionic_strength, Hc, Nc, Cc]).T, header='temperature ionic_strength, His_charge, N_term_charge, C_term_charge')\n",
    "\n",
    "# Need to store metadata prior to condacolab restarting the kernel\n",
    "f = open('seq.fasta','w')\n",
    "f.write('>{:s}\\n{:s}'.format(NAME,SEQUENCE))\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "MI2ZYXTzXLy6"
   },
   "outputs": [],
   "source": [
    "#@title <b>Preliminary operations</b>: setting the environment (i)\n",
    "import subprocess\n",
    "subprocess.run( 'pip install -q condacolab'.split() )\n",
    "import condacolab\n",
    "condacolab.install()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "pchP56o5XSGB"
   },
   "outputs": [],
   "source": [
    "#@title <b>Preliminary operations</b>: setting the environment (ii)\n",
    "import subprocess\n",
    "print('Installing libraries...')\n",
    "_ = subprocess.run( 'mamba install matplotlib mdtraj openmm=7.7 -c conda-forge --yes'.split() )\n",
    "subprocess.run( 'pip install wget localcider==0.1.18'.split() )\n",
    "import wget\n",
    "wget.download('https://raw.githubusercontent.com/fpesceKU/BLOCKING/main/block_tools.py')\n",
    "wget.download('https://raw.githubusercontent.com/fpesceKU/BLOCKING/main/main.py')\n",
    "import os\n",
    "import shutil\n",
    "import itertools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as scs\n",
    "from scipy.optimize import curve_fit\n",
    "from sklearn.covariance import LedoitWolf\n",
    "from scipy import constants\n",
    "from numpy import linalg\n",
    "from localcider.sequenceParameters import SequenceParameters\n",
    "from main import BlockAnalysis\n",
    "import mdtraj as md\n",
    "from simtk import openmm, unit\n",
    "from simtk.openmm import app\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from google.colab import files\n",
    "import matplotlib_inline.backend_inline\n",
    "matplotlib_inline.backend_inline.set_matplotlib_formats('pdf', 'svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "O_csIWvpiHht"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#A79AB2'>MD Toolbox</font></b>\n",
    "url = 'https://github.com/KULL-Centre/_2023_Tesei_IDRome/blob/main'\n",
    "wget.download(url+'/md_simulations/data/residues.csv?raw=true')\n",
    "\n",
    "residues = pd.read_csv('residues.csv')\n",
    "residues = residues.set_index('one')\n",
    "\n",
    "def genParamsLJ(df,seq,Nc,Cc):\n",
    "    fasta = seq.copy()\n",
    "    r = df.copy()\n",
    "    if Nc == 1:\n",
    "        r.loc['X'] = r.loc[fasta[0]]\n",
    "        r.loc['X','MW'] += 2\n",
    "    if Cc == 1:\n",
    "        r.loc['Z'] = r.loc[fasta[-1]]\n",
    "        r.loc['Z','MW'] += 16\n",
    "    lj_eps = 0.2*4.184\n",
    "    lj_sigma = pd.DataFrame((r.sigmas.values+r.sigmas.values.reshape(-1,1))/2,\n",
    "                            index=r.sigmas.index,columns=r.sigmas.index)\n",
    "    lj_lambda = pd.DataFrame((r.lambdas.values+r.lambdas.values.reshape(-1,1))/2,\n",
    "                             index=r.lambdas.index,columns=r.lambdas.index)\n",
    "    return lj_eps, lj_sigma, lj_lambda\n",
    "\n",
    "def genParamsDH(df,seq,temp,ionic,Nc,Cc,Hc):\n",
    "    kT = 8.3145*temp*1e-3\n",
    "    fasta = seq.copy()\n",
    "    r = df.copy()\n",
    "    # Set the charge on HIS based on the pH of the protein solution\n",
    "    r.loc['H','q'] = Hc\n",
    "    if Nc == 1:\n",
    "        r.loc['X'] = r.loc[fasta[0]]\n",
    "        r.loc['X','q'] = r.loc[seq[0],'q'] + 1.\n",
    "        fasta[0] = 'X'\n",
    "    if Cc == 1:\n",
    "        r.loc['Z'] = r.loc[fasta[-1]]\n",
    "        r.loc['Z','q'] = r.loc[seq[-1],'q'] - 1.\n",
    "        fasta[-1] = 'Z'\n",
    "    # Calculate the prefactor for the Yukawa potential\n",
    "    fepsw = lambda T : 5321/T+233.76-0.9297*T+0.1417*1e-2*T*T-0.8292*1e-6*T**3\n",
    "    epsw = fepsw(temp)\n",
    "    lB = 1.6021766**2/(4*np.pi*8.854188*epsw)*6.022*1000/kT\n",
    "    yukawa_eps = [r.loc[a].q*np.sqrt(lB*kT) for a in fasta]\n",
    "    # Calculate the inverse of the Debye length\n",
    "    yukawa_kappa = np.sqrt(8*np.pi*lB*ionic*6.022/10)\n",
    "    return yukawa_eps, yukawa_kappa\n",
    "\n",
    "def genDCD(name, seq, eqsteps=10):\n",
    "    \"\"\"\n",
    "    Generates coordinate and trajectory\n",
    "    in convenient formats\n",
    "    \"\"\"\n",
    "    traj = md.load(\"{:s}/pretraj.dcd\".format(name), top=\"{:s}/top.pdb\".format(name))\n",
    "    cgtop = md.Topology()\n",
    "    cgchain = cgtop.add_chain()\n",
    "    for aa in seq:\n",
    "        cgres = cgtop.add_residue(aa, cgchain)\n",
    "        cgtop.add_atom('CA', element=md.element.carbon, residue=cgres)\n",
    "    for i in range(traj.n_atoms-1):\n",
    "        cgtop.add_bond(cgtop.atom(i),cgtop.atom(i+1))\n",
    "    traj = md.Trajectory(traj.xyz, cgtop, traj.time, traj.unitcell_lengths, traj.unitcell_angles)\n",
    "    traj = traj.image_molecules(inplace=False, anchor_molecules=[set(traj.top.chain(0).atoms)], make_whole=True)\n",
    "    traj.center_coordinates()\n",
    "    traj.xyz += traj.unitcell_lengths[0,0]/2\n",
    "    traj[int(eqsteps):].save_dcd(\"{:s}/traj.dcd\".format(name))\n",
    "    traj[int(eqsteps)].save_pdb(\"{:s}/top.pdb\".format(name))\n",
    "\n",
    "def simulate(residues,name,seq,temp,ionic,Nc,Cc,Hc,nsteps,stride=1e3,eqsteps=1000):\n",
    "    os.mkdir(name)\n",
    "\n",
    "    lj_eps, _, _ = genParamsLJ(residues,seq,Nc,Cc)\n",
    "    yukawa_eps, yukawa_kappa = genParamsDH(residues,seq,temp,ionic,Nc,Cc,Hc)\n",
    "\n",
    "    N = len(seq)\n",
    "    L = (N-1)*0.38+4\n",
    "\n",
    "    system = openmm.System()\n",
    "\n",
    "    # set box vectors\n",
    "    a = unit.Quantity(np.zeros([3]), unit.nanometers)\n",
    "    a[0] = L * unit.nanometers\n",
    "    b = unit.Quantity(np.zeros([3]), unit.nanometers)\n",
    "    b[1] = L * unit.nanometers\n",
    "    c = unit.Quantity(np.zeros([3]), unit.nanometers)\n",
    "    c[2] = L * unit.nanometers\n",
    "    system.setDefaultPeriodicBoxVectors(a, b, c)\n",
    "\n",
    "    top = md.Topology()\n",
    "    pos = []\n",
    "    chain = top.add_chain()\n",
    "    pos.append([[0,0,L/2+(i-N/2.)*.38] for i in range(N)])\n",
    "    for resname in seq:\n",
    "        residue = top.add_residue(resname, chain)\n",
    "        top.add_atom(resname, element=md.element.carbon, residue=residue)\n",
    "    for i in range(chain.n_atoms-1):\n",
    "        top.add_bond(chain.atom(i),chain.atom(i+1))\n",
    "    md.Trajectory(np.array(pos).reshape(N,3), top, 0, [L,L,L], [90,90,90]).save_pdb('{:s}/top.pdb'.format(name))\n",
    "\n",
    "    pdb = app.pdbfile.PDBFile('{:s}/top.pdb'.format(name))\n",
    "\n",
    "    system.addParticle((residues.loc[seq[0]].MW+2)*unit.amu)\n",
    "    for a in seq[1:-1]:\n",
    "        system.addParticle(residues.loc[a].MW*unit.amu)\n",
    "    system.addParticle((residues.loc[seq[-1]].MW+16)*unit.amu)\n",
    "\n",
    "    hb = openmm.openmm.HarmonicBondForce()\n",
    "    energy_expression = 'select(step(r-2^(1/6)*s),4*eps*l*((s/r)^12-(s/r)^6-shift),4*eps*((s/r)^12-(s/r)^6-l*shift)+eps*(1-l))'\n",
    "    ah = openmm.openmm.CustomNonbondedForce(energy_expression+'; s=0.5*(s1+s2); l=0.5*(l1+l2); shift=(0.5*(s1+s2)/2)^12-(0.5*(s1+s2)/2)^6')\n",
    "    yu = openmm.openmm.CustomNonbondedForce('q*(exp(-kappa*r)/r - exp(-kappa*4)/4); q=q1*q2')\n",
    "    yu.addGlobalParameter('kappa',yukawa_kappa/unit.nanometer)\n",
    "    yu.addPerParticleParameter('q')\n",
    "\n",
    "    ah.addGlobalParameter('eps',lj_eps*unit.kilojoules_per_mole)\n",
    "    ah.addPerParticleParameter('s')\n",
    "    ah.addPerParticleParameter('l')\n",
    "\n",
    "    for a,e in zip(seq,yukawa_eps):\n",
    "        yu.addParticle([e*unit.nanometer*unit.kilojoules_per_mole])\n",
    "        ah.addParticle([residues.loc[a].sigmas*unit.nanometer, residues.loc[a].lambdas*unit.dimensionless])\n",
    "\n",
    "    for i in range(N-1):\n",
    "        hb.addBond(i, i+1, 0.38*unit.nanometer, 8033*unit.kilojoules_per_mole/(unit.nanometer**2))\n",
    "        yu.addExclusion(i, i+1)\n",
    "        ah.addExclusion(i, i+1)\n",
    "\n",
    "    yu.setForceGroup(0)\n",
    "    ah.setForceGroup(1)\n",
    "    yu.setNonbondedMethod(openmm.openmm.CustomNonbondedForce.CutoffPeriodic)\n",
    "    ah.setNonbondedMethod(openmm.openmm.CustomNonbondedForce.CutoffPeriodic)\n",
    "    hb.setUsesPeriodicBoundaryConditions(True)\n",
    "    yu.setCutoffDistance(4*unit.nanometer)\n",
    "    ah.setCutoffDistance(2*unit.nanometer)\n",
    "\n",
    "    system.addForce(hb)\n",
    "    system.addForce(yu)\n",
    "    system.addForce(ah)\n",
    "\n",
    "    #serialized_system = XmlSerializer.serialize(system)\n",
    "    #outfile = open('system.xml','w')\n",
    "    #outfile.write(serialized_system)\n",
    "    #outfile.close()\n",
    "\n",
    "    integrator = openmm.openmm.LangevinIntegrator(temp*unit.kelvin,0.01/unit.picosecond,0.010*unit.picosecond) #10 fs timestep\n",
    "\n",
    "    platform = openmm.Platform.getPlatformByName('CUDA')\n",
    "\n",
    "    simulation = app.simulation.Simulation(pdb.topology, system, integrator, platform, dict(CudaPrecision='mixed'))\n",
    "\n",
    "    check_point = '{:s}/restart.chk'.format(name)\n",
    "\n",
    "    if os.path.isfile(check_point):\n",
    "        print('Reading check point file')\n",
    "        simulation.loadCheckpoint(check_point)\n",
    "        simulation.reporters.append(app.dcdreporter.DCDReporter('{:s}/pretraj.dcd'.format(name),int(stride),append=True))\n",
    "    else:\n",
    "        simulation.context.setPositions(pdb.positions)\n",
    "        simulation.minimizeEnergy()\n",
    "        simulation.reporters.append(app.dcdreporter.DCDReporter('{:s}/pretraj.dcd'.format(name),int(stride)))\n",
    "\n",
    "    simulation.reporters.append(app.statedatareporter.StateDataReporter('{:s}/traj.log'.format(name),int(stride),\n",
    "             potentialEnergy=True,temperature=True,step=True,speed=True,elapsedTime=True,separator='\\t'))\n",
    "\n",
    "    simulation.step(nsteps)\n",
    "\n",
    "    simulation.saveCheckpoint(check_point)\n",
    "\n",
    "    genDCD(name,[residues.three[aa] for aa in seq],eqsteps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "6Y3hzuNUk0rd"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#A79AB2'>Sequence analysis Toolbox</font></b>\n",
    "def calc_seq_prop(seq,residues,Nc,Cc,Hc):\n",
    "    \"\"\"df: DataFrame to be populated with sequence properties\n",
    "    r: DataFrame of aa-specific parameters\"\"\"\n",
    "    seq = list(seq).copy()\n",
    "    fasta_kappa = np.array(seq.copy())\n",
    "    N = len(seq)\n",
    "    r = residues.copy()\n",
    "\n",
    "    # calculate properties that do not depend on charges\n",
    "    fK = sum([seq.count(a) for a in ['K']])/N\n",
    "    fR = sum([seq.count(a) for a in ['R']])/N\n",
    "    fE = sum([seq.count(a) for a in ['E']])/N\n",
    "    fD = sum([seq.count(a) for a in ['D']])/N\n",
    "    faro = sum([seq.count(a) for a in ['W','Y','F']])/N\n",
    "    mean_lambda = np.mean(r.loc[seq].lambdas)\n",
    "\n",
    "    pairs = np.array(list(itertools.combinations(seq,2)))\n",
    "    pairs_indices = np.array(list(itertools.combinations(range(N),2)))\n",
    "    # calculate sequence separations\n",
    "    ij_dist = np.diff(pairs_indices,axis=1).flatten().astype(float)\n",
    "    # calculate lambda sums\n",
    "    ll = r.lambdas.loc[pairs[:,0]].values+r.lambdas.loc[pairs[:,1]].values\n",
    "    # calculate SHD\n",
    "    beta = -1\n",
    "    shd = np.sum(ll*np.power(np.abs(ij_dist),beta))/N\n",
    "\n",
    "    # fix charges\n",
    "    if Nc == 1:\n",
    "        r.loc['X'] = r.loc[seq[0]]\n",
    "        r.loc['X','q'] = r.loc[seq[0],'q'] + 1.\n",
    "        seq[0] = 'X'\n",
    "        if r.loc['X','q'] > 0:\n",
    "            fasta_kappa[0] = 'K'\n",
    "        else:\n",
    "            fasta_kappa[0] = 'A'\n",
    "    if Cc == 1:\n",
    "        r.loc['Z'] = r.loc[seq[-1]]\n",
    "        r.loc['Z','q'] = r.loc[seq[-1],'q'] - 1.\n",
    "        seq[-1] = 'Z'\n",
    "        if r.loc['Z','q'] < 0:\n",
    "            fasta_kappa[-1] = 'D'\n",
    "        else:\n",
    "            fasta_kappa[-1] = 'A'\n",
    "    if Hc < 0.5:\n",
    "        r.loc['H', 'q'] = 0\n",
    "        fasta_kappa[np.where(np.array(seq) == 'H')[0]] = 'A'\n",
    "    elif Hc >= 0.5:\n",
    "        r.loc['H', 'q'] = 1\n",
    "        fasta_kappa[np.where(np.array(seq) == 'H')[0]] = 'K'\n",
    "\n",
    "    # calculate properties that depend on charges\n",
    "    pairs = np.array(list(itertools.combinations(seq,2)))\n",
    "    # calculate charge products\n",
    "    qq = r.q.loc[pairs[:,0]].values*r.q.loc[pairs[:,1]].values\n",
    "    # calculate SCD\n",
    "    scd = np.sum(qq*np.sqrt(ij_dist))/N\n",
    "    SeqOb = SequenceParameters(''.join(fasta_kappa))\n",
    "    kappa = SeqOb.get_kappa()\n",
    "    fcr = r.q.loc[seq].abs().mean()\n",
    "    ncpr = r.q.loc[seq].mean()\n",
    "\n",
    "    return np.around([fK, fR, fE, fD, faro, mean_lambda, shd, scd, kappa, fcr, ncpr],3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "uf4w32WliiFK"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#A79AB2'>Simulation analysis Toolbox</font></b>\n",
    "\n",
    "def autoblock(cv, multi=1, plot=False):\n",
    "    block = BlockAnalysis(cv, multi=multi)\n",
    "    block.SEM()\n",
    "\n",
    "    if plot == True:\n",
    "        plt.errorbar(block.stat[...,0], block.stat[...,1], block.stat[...,2], fmt='', color='k', ecolor='0.5')\n",
    "        plt.scatter(block.bs, block.sem,zorder=10,c='tab:red')\n",
    "        plt.xlabel('Block size')\n",
    "        plt.ylabel('SEM')\n",
    "        plt.show()\n",
    "\n",
    "    return block.av, block.sem, block.bs\n",
    "\n",
    "def error_ratio(v1,v2,e1,e2):\n",
    "    ratio = v1/v2\n",
    "    return ratio*np.sqrt((e1/v1)**2+(e2/v2)**2)\n",
    "\n",
    "def calc_shape_and_entropy(t,forces,residues,seq,temp):\n",
    "    fasta = list(seq)\n",
    "    masses = residues.loc[fasta,'MW'].values[np.newaxis,:,np.newaxis]\n",
    "\n",
    "    # calculate conformational entropy per residue (DOI: 10.1021/ct500684w)\n",
    "    prefactor = constants.Boltzmann*temp/constants.hbar**2\n",
    "    forces = forces/np.sqrt(masses/1e3/constants.Avogadro)\n",
    "    forces = forces.reshape(t.n_frames,-1)*1e3/constants.Avogadro*1e9\n",
    "    sigma = LedoitWolf().fit(forces).covariance_\n",
    "    eigenvalues = linalg.eigvalsh(sigma)\n",
    "    kT_over_hbar_omega = constants.Boltzmann*temp*np.sqrt(prefactor/eigenvalues)\n",
    "    SPR = np.sum(np.log(kT_over_hbar_omega+1))/len(seq) # R\n",
    "\n",
    "    # calculate the center of mass\n",
    "    cm = np.sum(t.xyz*masses,axis=1)/masses.sum()\n",
    "    # calculate residue-cm distances\n",
    "    si = t.xyz - cm[:,np.newaxis,:]\n",
    "    q = np.einsum('jim,jin->jmn', si*masses,si)/masses.sum()\n",
    "    trace_q = np.trace(q,axis1=1,axis2=2)\n",
    "    # calculate rg\n",
    "    rgarray = np.sqrt(trace_q)\n",
    "    # calculate traceless matrix\n",
    "    mean_trace = np.trace(q,axis1=1,axis2=2)/3\n",
    "    q_hat = q - mean_trace.reshape(-1,1,1)*np.identity(3).reshape(-1,3,3)\n",
    "    # calculate asphericity\n",
    "    Delta_array = 3/2*np.trace(q_hat**2,axis1=1,axis2=2)/(trace_q**2)\n",
    "    # calculate oblateness\n",
    "    S_array = 27*linalg.det(q_hat)/(trace_q**3)\n",
    "    # calculate ensemble averages\n",
    "    block_tr_q_hat_2 = BlockAnalysis(np.trace(q_hat**2,axis1=1,axis2=2), multi=1)\n",
    "    block_tr_q_hat_2.SEM()\n",
    "    block_tr_q_2 = BlockAnalysis(trace_q**2, multi=1)\n",
    "    block_tr_q_2.SEM()\n",
    "    block_det_q_hat = BlockAnalysis(linalg.det(q_hat), multi=1)\n",
    "    block_det_q_hat.SEM()\n",
    "    block_tr_q_3 = BlockAnalysis(trace_q**3, multi=1)\n",
    "    block_tr_q_3.SEM()\n",
    "    Delta = 3/2*block_tr_q_hat_2.av/block_tr_q_2.av\n",
    "    S = 27*block_det_q_hat.av/block_tr_q_3.av\n",
    "    Delta_err = 3/2*error_ratio(block_tr_q_hat_2.av,block_tr_q_2.av,block_tr_q_hat_2.sem,block_tr_q_2.sem)\n",
    "    S_err = 27*error_ratio(block_det_q_hat.av,block_tr_q_3.av,block_det_q_hat.sem,block_tr_q_3.sem)\n",
    "    return rgarray, Delta_array, S_array, Delta, S, Delta_err, S_err, SPR\n",
    "\n",
    "def Rij(traj):\n",
    "    pairs = traj.top.select_pairs('all','all')\n",
    "    d = md.compute_distances(traj,pairs)\n",
    "    nres = traj.n_atoms\n",
    "    ij = np.arange(2,nres,1)\n",
    "    diff = [x[1]-x[0] for x in pairs]\n",
    "    dij = np.empty(0)\n",
    "    for i in ij:\n",
    "        dij = np.append(dij,np.sqrt((d[:,diff==i]**2).mean().mean()))\n",
    "    f = lambda x,R0,v : R0*np.power(x,v)\n",
    "    popt, pcov = curve_fit(f,ij[ij>5],dij[ij>5],p0=[.4,.5])\n",
    "    nu = popt[1]\n",
    "    nu_err = pcov[1,1]**0.5\n",
    "    R0 = popt[0]\n",
    "    R0_err = pcov[0,0]**0.5\n",
    "    return ij,dij,nu,nu_err,R0,R0_err\n",
    "\n",
    "#energy functions\n",
    "HALR = lambda r,s,l : 4*0.8368*l*((s/r)**12-(s/r)**6)\n",
    "HASR = lambda r,s,l : 4*0.8368*((s/r)**12-(s/r)**6)+0.8368*(1-l)\n",
    "HA = lambda r,s,l : np.where(r<2**(1/6)*s, HASR(r,s,l), HALR(r,s,l))\n",
    "HASP = lambda r,s,l,rc : np.where(r<rc, HA(r,s,l)-HA(rc,s,l), 0)\n",
    "\n",
    "#force functions\n",
    "LJ_F = lambda r,s,rvec : -6*4*0.8368*(2*s**12/r**14-s**6/r**8)*rvec\n",
    "HA_F = lambda r,s,l,rvec : np.where(r<2**(1/6)*s, LJ_F(r,s,rvec), l*LJ_F(r,s,rvec))\n",
    "HASP_F = lambda r,s,l,rvec,rc : np.where(r<rc, HA_F(r,s,l,rvec), 0)\n",
    "\n",
    "DH_F = lambda r,yukawa_eps,yukawa_kappa,rvec : -yukawa_eps*np.exp(-r*yukawa_kappa)*(1+r*yukawa_kappa)/r**3*rvec\n",
    "DHSP_F = lambda r,yukawa_eps,yukawa_kappa,rvec,rc : np.where(r<rc, DH_F(r,yukawa_eps,yukawa_kappa,rvec), 0)\n",
    "\n",
    "def calc_energy_map(t,df,seq,rc,temp,ionic):\n",
    "    indices = t.top.select_pairs('all','all')\n",
    "    mask = np.abs(indices[:,0]-indices[:,1])>1 #exclude bonded pairs\n",
    "    indices = indices[mask]\n",
    "    dvec = md.compute_displacements(t,indices) #vector distances between pairs for each frame\n",
    "    d = linalg.norm(dvec,axis=2)\n",
    "    pairs = np.array(list(itertools.combinations(list(seq),2)))\n",
    "    pairs = pairs[mask]\n",
    "    sigmas = 0.5*(df.loc[pairs[:,0]].sigmas.values+df.loc[pairs[:,1]].sigmas.values)\n",
    "    lambdas = 0.5*(df.loc[pairs[:,0]].lambdas.values+df.loc[pairs[:,1]].lambdas.values)\n",
    "    RT = 8.3145*temp*1e-3\n",
    "    fepsw = lambda T : 5321/T+233.76-0.9297*T+0.1417*1e-2*T*T-0.8292*1e-6*T**3\n",
    "    epsw = fepsw(temp)\n",
    "    lB = 1.6021766**2/(4*np.pi*8.854188*epsw)*6.022*1000/RT\n",
    "    # Calculate the inverse of the Debye length\n",
    "    yukawa_kappa = np.sqrt(8*np.pi*lB*ionic*6.022/10)\n",
    "    qq = df.loc[pairs[:,0]].q.values*df.loc[pairs[:,1]].q.values\n",
    "    yukawa_eps = qq*lB*RT\n",
    "    emap = np.zeros(pairs.shape[0])\n",
    "    forces = np.zeros((t.n_frames,t.n_atoms,3))\n",
    "    dstd = np.std(d,axis=0)\n",
    "    for j,(r,rvec) in enumerate(zip(np.split(d,20,axis=0),np.split(dvec,20,axis=0))):\n",
    "        emap += np.nansum(HASP(r,sigmas[np.newaxis,:],lambdas[np.newaxis,:],rc),axis=0)\n",
    "        for i in range(t.n_atoms):\n",
    "            ndx_pairs = np.any(indices==i,axis=1)\n",
    "            forces[j*r.shape[0]:(j+1)*r.shape[0],i] = np.nansum(\n",
    "                HASP_F(r[:,ndx_pairs,np.newaxis],sigmas[np.newaxis,ndx_pairs,np.newaxis],\n",
    "                            lambdas[np.newaxis,ndx_pairs,np.newaxis],rvec[:,ndx_pairs],rc),axis=1)\n",
    "            forces[j*r.shape[0]:(j+1)*r.shape[0],i] += np.nansum(DHSP_F(r[:,ndx_pairs,np.newaxis],\n",
    "                            yukawa_eps[np.newaxis,ndx_pairs,np.newaxis],yukawa_kappa,rvec[:,ndx_pairs],4),axis=1)\n",
    "    return indices, emap/d.shape[0], forces, dstd\n",
    "\n",
    "def Ree(t):\n",
    "    return md.compute_distances( t, atom_pairs=np.array([[ 0,  len(list(t.top.atoms))-1]]) )[...,0]\n",
    "\n",
    "def maps(traj,residues,seq,temp,ionic):\n",
    "    #energy maps\n",
    "    df_emap = pd.DataFrame(index=range(traj.n_atoms),columns=range(traj.n_atoms),dtype=float)\n",
    "    df_dstd = pd.DataFrame(index=range(traj.n_atoms),columns=range(traj.n_atoms),dtype=float)\n",
    "    pairs, emap, forces, dstd = calc_energy_map(traj,residues,seq,2.0,temp,ionic)\n",
    "    for k,(i,j) in enumerate(pairs):\n",
    "        df_emap.loc[i,j] = emap[k]\n",
    "        df_emap.loc[j,i] = emap[k]\n",
    "        df_dstd.loc[i,j] = dstd[k]\n",
    "        df_dstd.loc[j,i] = dstd[k]\n",
    "    return df_emap, forces, df_dstd\n",
    "\n",
    "def kde(a):\n",
    "    min_ = np.min(a)\n",
    "    max_ = np.max(a)\n",
    "    x = np.linspace( min_, max_, num = 100 )\n",
    "    d = scs.gaussian_kde( a, bw_method = \"silverman\" ).evaluate(x)\n",
    "    u = np.average(a)\n",
    "    return x,d/np.sum(d),u"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "pVgiFPDoqyz4"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#F26419'>2 - Sequence analysis</font></b>\n",
    "df = pd.DataFrame(columns=['fK','fR','fE','fD','fARO','mean_lambda','SHD','SCD','kappa','FCR','NCPR'])\n",
    "f = open('seq.fasta', 'r').readlines()\n",
    "NAME = f[0][1:].strip()\n",
    "SEQUENCE = f[1].strip()\n",
    "Temperature, Ionic_strength, Hc, Nc, Cc = np.loadtxt('env_settings.txt', unpack=True)\n",
    "\n",
    "df.loc[NAME] = calc_seq_prop(SEQUENCE,residues,Nc,Cc,Hc)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "KDf4Vso9sLq-"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#45B69C'>3.1 - Run MD simulation</font></b>\n",
    "#@markdown Simulation time (ns):\n",
    "Simulation_time = \"AUTO\" #@param {type:\"raw\"}\n",
    "\n",
    "N_res = len(SEQUENCE)\n",
    "N_save = 7000 if N_res < 150 else int(np.ceil(3e-4*N_res**2)*1000)\n",
    "\n",
    "if Simulation_time == \"AUTO\":\n",
    "    nsteps = 1010*N_save\n",
    "    print('AUTO simulation length selected. Running for {} ns'.format(nsteps*0.01/1000))\n",
    "else:\n",
    "    nsteps = float(Simulation_time)*1000/0.01//N_save*N_save\n",
    "try:\n",
    "    shutil.rmtree(NAME)\n",
    "except:\n",
    "    pass\n",
    "simulate(residues,NAME,list(SEQUENCE),temp=Temperature,ionic=Ionic_strength,Nc=Nc,Cc=Cc,\n",
    "         Hc=Hc,nsteps=nsteps,stride=N_save,eqsteps=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "wpdRnLUnHkyb"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#45B69C'>3.2 - Simulation analysis</font></b>\n",
    "traj = md.load_dcd('{:s}/traj.dcd'.format(NAME), top='{:s}/top.pdb'.format(NAME))\n",
    "\n",
    "df_emap, forces, df_dstd = maps(traj,residues,SEQUENCE,Temperature,Ionic_strength)\n",
    "\n",
    "rg_array, D_array, S_array, D, S, D_err, S_err, SPR = calc_shape_and_entropy(traj,forces,\n",
    "                                                                    residues,SEQUENCE,Temperature)\n",
    "\n",
    "rg, rg_err, _ = autoblock(rg_array)\n",
    "rg_hist = kde(rg_array)\n",
    "D_hist = kde(D_array)\n",
    "S_hist = kde(S_array)\n",
    "\n",
    "ree_array = Ree(traj)\n",
    "ree, ree_err, _ = autoblock(ree_array)\n",
    "ree_hist = kde(ree_array)\n",
    "\n",
    "ij,dij,nu,nu_err,R0,R0_err = Rij(traj)\n",
    "\n",
    "# Plot results\n",
    "mpl.rcParams.update({'font.size': 10})\n",
    "fig, axs = plt.subplots(2, 3, figsize=(7,3.5), facecolor='w', dpi=300, layout='tight')\n",
    "axs = axs.flatten()\n",
    "\n",
    "axs[0].plot(rg_hist[0],rg_hist[1])\n",
    "top = rg_hist[1].max()+0.1*rg_hist[1].max()\n",
    "axs[0].vlines(rg,0,top)\n",
    "axs[0].set_xlabel(r'$R_g$ (nm)')\n",
    "axs[0].set_ylabel(r'$p(R_g)$')\n",
    "axs[0].set_ylim(0,top)\n",
    "axs[0].fill_between([rg-rg_err,rg+rg_err],0,top,alpha=0.3)\n",
    "\n",
    "axs[1].plot(D_hist[0],D_hist[1])\n",
    "top = D_hist[1].max()+0.1*D_hist[1].max()\n",
    "axs[1].vlines(D,0,top)\n",
    "axs[1].set_xlabel(r'$\\Delta$')\n",
    "axs[1].set_ylabel(r'$p(\\Delta)$')\n",
    "axs[1].set_ylim(0,top)\n",
    "axs[1].fill_between([D-D_err,D+D_err],0,top,alpha=0.3)\n",
    "\n",
    "axs[2].plot(S_hist[0],S_hist[1])\n",
    "top = S_hist[1].max()+0.1*S_hist[1].max()\n",
    "axs[2].vlines(S,0,top)\n",
    "axs[2].set_xlabel(r'$S$')\n",
    "axs[2].set_ylabel(r'$p(S)$')\n",
    "axs[2].set_ylim(0,top)\n",
    "axs[2].fill_between([S-S_err,S+S_err],0,top,alpha=0.3)\n",
    "\n",
    "axs[3].plot(ree_hist[0],ree_hist[1])\n",
    "top = ree_hist[1].max()+0.1*ree_hist[1].max()\n",
    "axs[3].vlines(ree,0,top)\n",
    "axs[3].set_xlabel(r'$R_{ee}$ (nm)')\n",
    "axs[3].set_ylabel(r'$p(R_{ee})$')\n",
    "axs[3].set_ylim(0,top)\n",
    "\n",
    "axs[4].plot(ij,dij)\n",
    "dij_fit = R0*np.power(ij,nu)\n",
    "axs[4].plot(ij, dij_fit,c='0.3',ls='dashed',label='Fit')\n",
    "axs[4].set_xlabel('$|i-j|$')\n",
    "axs[4].set_ylabel(r'$\\sqrt{\\langle R_{ij}^2 \\rangle}$ (nm)')\n",
    "axs[4].text(0.05, 0.9, r'$\\nu$={:.2f}'.format(nu), horizontalalignment='left',\n",
    "            verticalalignment='center', transform=axs[4].transAxes, fontsize=10)\n",
    "axs[4].legend(loc='lower right')\n",
    "\n",
    "im = axs[5].imshow(df_emap*1e3,extent=[1, df_emap.shape[0], 1, df_emap.shape[0]],origin='lower',\n",
    "                   aspect='equal',vmin=-9,vmax=0,cmap=plt.cm.Blues_r)\n",
    "cb = plt.colorbar(im, ax=axs[5], label=r'$U$ (J mol$^{-1}$)',fraction=0.05, pad=0.04)\n",
    "cb.set_ticks([0,-3,-6,-9])\n",
    "axs[5].set_xlabel('Residue #')\n",
    "axs[5].set_ylabel('Residue #')\n",
    "\n",
    "plt.savefig('conformational_properties.pdf', dpi=300, facecolor='w', edgecolor='w', orientation='portrait',\n",
    "            bbox_inches='tight')\n",
    "plt.show()\n",
    "\n",
    "df_means = pd.DataFrame(data=np.c_[[rg,rg_err],[ree,ree_err],[nu,nu_err],[D,D_err],[S,S_err],[SPR,0]],\n",
    "                        columns=['<Rg> (nm)','<Ree> (nm)','nu','<Delta>','<S>','Sconf/N (kB)'],\n",
    "                        index=['Value','Error'])\n",
    "df_means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "-SCC7r8ltUTy"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#E3B505'>4 - Download results</font></b>\n",
    "try:\n",
    "    os.remove('{}/pretraj.dcd'.format(NAME))\n",
    "    os.remove('{}/restart.chk'.format(NAME))\n",
    "except:\n",
    "    pass\n",
    "shutil.copy('env_settings.txt','{}/env_settings.txt'.format(NAME))\n",
    "try:\n",
    "    os.mkdir('{}_IDRLab'.format(NAME))\n",
    "except:\n",
    "    pass\n",
    "shutil.copytree('{:s}'.format(NAME), '{:s}_IDRLab/SIMULATION'.format(NAME),dirs_exist_ok = True)\n",
    "try:\n",
    "    shutil.move('conformational_properties.pdf', '{:s}_IDRLab/conformational_properties.pdf'.format(NAME))\n",
    "except:\n",
    "    pass\n",
    "\n",
    "pd.DataFrame(data=np.c_[rg_array,ree_array,D_array,S_array],\n",
    "             columns=['Rg (nm)','Ree (nm)','Delta','S']).to_csv(\n",
    "            '{:s}_IDRLab/time_series_Rg_Ree_Delta_S.csv'.format(NAME))\n",
    "pd.DataFrame(data=np.c_[ij,dij,dij_fit],\n",
    "             columns=['ij','Rij (nm)','Rij_fit (nm)']).to_csv('{:s}_IDRLab/scaling_profile.csv'.format(NAME))\n",
    "df.to_csv('{:s}_IDRLab/sequence_properties.csv'.format(NAME))\n",
    "df_means.to_csv('{:s}_IDRLab/mean_structural_parameters.csv'.format(NAME))\n",
    "df_emap.to_csv('{:s}_IDRLab/energy_map.csv'.format(NAME))\n",
    "df_dstd.to_csv('{:s}_IDRLab/stdev_map.csv'.format(NAME))\n",
    "\n",
    "zipper = 'zip -r {:s}_IDRLab.zip {:s}_IDRLab'.format(NAME,NAME)\n",
    "subprocess.run(zipper.split())\n",
    "files.download('{:s}_IDRLab.zip'.format(NAME))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "GGj6wUOhIElY"
   },
   "source": [
    "For further analyses, the coarse-grained trajectory can be converted to an all-atom trajectory using the _powerful chain restoration algorithm_ (PULCHRA) [4].\n",
    "\n",
    "[4] P. Rotkiewicz and J. Skolnick __Fast procedure for reconstruction of full-atom protein models from reduced representations__ _J. Comput. Chem._ 2008 29:1460â€“1465 DOI: https://doi.org/10.1002/jcc.20906\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "EzjkbgzDG40_"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#A79AB2'>5.1 - Download and install PULCHRA</font></b>\n",
    "%%bash\n",
    "\n",
    "wget https://cssb.biology.gatech.edu/skolnick/files/PULCHRA/pulchra304.tgz &> /dev/null\n",
    "tar -zxf pulchra304.tgz &> /dev/null\n",
    "rm pulchra304.tgz\n",
    "mv ./pulchra304/bin/linux/pulchra .\n",
    "chmod +x pulchra\n",
    "rm -r pulchra304/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "tZTHMPzDG40_"
   },
   "outputs": [],
   "source": [
    "#@title <b><font color='#A79AB2'>5.2 - Convert to all-atom using PULCHRA</font></b>\n",
    "\n",
    "def backmapping(name, traj):\n",
    "    for i in range(traj.n_frames):\n",
    "        traj[i].save_pdb('frame.pdb')\n",
    "        subprocess.run(['./pulchra', 'frame.pdb'])\n",
    "        if i == 0:\n",
    "            traj_AA = md.load_pdb('frame.rebuilt.pdb')\n",
    "        else:\n",
    "            traj_AA += md.load_pdb('frame.rebuilt.pdb')\n",
    "    top = md.Topology()\n",
    "    chain = top.add_chain()\n",
    "    for residue in traj_AA.top.residues:\n",
    "        res = top.add_residue(residue.name, chain, resSeq=residue.index+1)\n",
    "        for atom in residue.atoms:\n",
    "            top.add_atom(atom.name, element=atom.element, residue=res)\n",
    "    traj_AA = md.Trajectory(traj_AA.xyz, top, traj.time, traj.unitcell_lengths, traj.unitcell_angles)\n",
    "    traj_AA[0].save_pdb(f'{name:s}/all_atom/top_AA.pdb')\n",
    "    traj_AA.save_dcd(f'{name:s}/all_atom/traj_AA.dcd')\n",
    "    return traj_AA\n",
    "\n",
    "THREE_LETTER_SEQ = [residues.three[x] for x in SEQUENCE]\n",
    "traj = md.load_dcd('{:s}/traj.dcd'.format(NAME), top='{:s}/top.pdb'.format(NAME))\n",
    "\n",
    "try:\n",
    "    os.makedirs(f'{NAME:s}/all_atom')\n",
    "except:\n",
    "    pass\n",
    "\n",
    "backmapping(NAME, traj)\n",
    "\n",
    "zipper = 'zip -r {:s}_all_atom.zip {:s}/all_atom'.format(NAME,NAME)\n",
    "subprocess.run(zipper.split())\n",
    "files.download('{:s}_all_atom.zip'.format(NAME))"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
