{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a href=\"https://colab.research.google.com/github/KULL-Centre/_2023_Tesei_IDRome/blob/main/nu_SVR_predictor.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Preliminary information:**\n",
    "\n",
    "This Colab notebook enables to predict the scaling exponent, $\\nu$, of an intrinsically disordered proteins (IDPs) or protein region (IDRs) based on the amino acid sequence.\n",
    "\n",
    "Predictions are generated by a support vector regression (SVR) model [1], which was trained on simulations of all the IDPs and IDRs of the human proteome performed using the CALVADOS model [2].\n",
    "\n",
    "Amino acid sequences can be provided via (i) a single fasta file containing one or several entries, (ii) multiple fasta files, or (iii) pasting each sequence in separate input text boxes.\n",
    "\n",
    "<b><font color='#FA003F'>How to cite this notebook:</font></b> If you use the $\\nu$ values generated by the SVR model, we ask you to cite Tesei, Trolle et al. [1]. \n",
    "1. G. Tesei, A. I. Trolle, N. Jonsson, J. Betz, F. Pesce, K. E. Johansson, K. Lindorff-Larsen __Conformational ensembles of the human intrinsically disordered proteome: Bridging chain compaction with function and sequence conservation__ _bioRxiv_ 2023 2023.05.08.539815 DOI: https://doi.org/10.1101/2023.05.08.539815\n",
    "2. G. Tesei and K. Lindorff-Larsen __Improved predictions of phase behaviour of intrinsically disordered proteins by tuning the interaction range [version 2; peer review: 2 approved]__ _Open Research Europe_ 2023 2(94) DOI: https://doi.org/10.12688/openreseurope.14967.2\n",
    "---\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "fGNr_8e1Ocv_"
   },
   "outputs": [],
   "source": [
    "#@title <b>Preliminary operations</b>\n",
    "import subprocess\n",
    "subprocess.run( 'pip install wget localcider==0.1.18'.split() )\n",
    "subprocess.run('pip uninstall scikit-learn -y'.split())\n",
    "subprocess.run('pip install scikit-learn==1.0.2'.split())\n",
    "import numpy as np\n",
    "import itertools\n",
    "from localcider.sequenceParameters import SequenceParameters\n",
    "import wget\n",
    "import sys\n",
    "import os\n",
    "from joblib import dump, load\n",
    "import pandas as pd\n",
    "from google.colab import files\n",
    "from ipywidgets import IntProgress\n",
    "from IPython.display import display\n",
    "from IPython.display import clear_output\n",
    "\n",
    "def calc_seq_prop(seq,residues,Nc,Cc,Hc):\n",
    "    seq = list(seq).copy()\n",
    "    fasta_kappa = np.array(seq.copy())\n",
    "    N = len(seq)\n",
    "    r = residues.copy()\n",
    "\n",
    "    # calculate properties that do not depend on charges\n",
    "    fK = sum([seq.count(a) for a in ['K']])/N\n",
    "    fR = sum([seq.count(a) for a in ['R']])/N\n",
    "    fE = sum([seq.count(a) for a in ['E']])/N\n",
    "    fD = sum([seq.count(a) for a in ['D']])/N\n",
    "    faro = sum([seq.count(a) for a in ['W','Y','F']])/N\n",
    "    mean_lambda = np.mean(r.loc[seq].lambdas)\n",
    "\n",
    "    pairs = np.array(list(itertools.combinations(seq,2)))\n",
    "    pairs_indices = np.array(list(itertools.combinations(range(N),2)))\n",
    "    # calculate sequence separations\n",
    "    ij_dist = np.diff(pairs_indices,axis=1).flatten().astype(float)\n",
    "    # calculate lambda sums\n",
    "    ll = r.lambdas.loc[pairs[:,0]].values+r.lambdas.loc[pairs[:,1]].values\n",
    "    # calculate SHD\n",
    "    beta = -1\n",
    "    shd = np.sum(ll*np.power(np.abs(ij_dist),beta))/N\n",
    "    SeqOb = SequenceParameters(''.join(seq))\n",
    "    omega = SeqOb.get_kappa_X(grp1=['F','Y','W'])\n",
    "\n",
    "    # fix charges\n",
    "    if Nc == 1:\n",
    "        r.loc['X'] = r.loc[seq[0]]\n",
    "        r.loc['X','q'] = r.loc[seq[0],'q'] + 1.\n",
    "        seq[0] = 'X'\n",
    "        if r.loc['X','q'] > 0:\n",
    "            fasta_kappa[0] = 'K'\n",
    "        else:\n",
    "            fasta_kappa[0] = 'A'\n",
    "    if Cc == 1:\n",
    "        r.loc['Z'] = r.loc[seq[-1]]\n",
    "        r.loc['Z','q'] = r.loc[seq[-1],'q'] - 1.\n",
    "        seq[-1] = 'Z'\n",
    "        if r.loc['Z','q'] < 0:\n",
    "            fasta_kappa[-1] = 'D'\n",
    "        else:\n",
    "            fasta_kappa[-1] = 'A'\n",
    "    if Hc < 0.5:\n",
    "        r.loc['H', 'q'] = 0\n",
    "        fasta_kappa[np.where(np.array(seq) == 'H')[0]] = 'A'\n",
    "    elif Hc >= 0.5:\n",
    "        r.loc['H', 'q'] = 1\n",
    "        fasta_kappa[np.where(np.array(seq) == 'H')[0]] = 'K'\n",
    "\n",
    "    # calculate properties that depend on charges\n",
    "    pairs = np.array(list(itertools.combinations(seq,2)))\n",
    "    # calculate charge products\n",
    "    qq = r.q.loc[pairs[:,0]].values*r.q.loc[pairs[:,1]].values\n",
    "    # calculate SCD\n",
    "    scd = np.sum(qq*np.sqrt(ij_dist))/N\n",
    "    SeqOb = SequenceParameters(''.join(fasta_kappa))\n",
    "    kappa = SeqOb.get_kappa()\n",
    "    fcr = r.q.loc[seq].abs().mean()\n",
    "    ncpr = r.q.loc[seq].mean()\n",
    "\n",
    "    return np.around([fK, fR, fE, fD, faro, mean_lambda, shd, omega, scd, kappa, fcr, ncpr],3)\n",
    "\n",
    "aa = ['A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y']\n",
    "\n",
    "if os.path.exists('svr_model.joblib') == False:\n",
    "    wget.download('https://github.com/KULL-Centre/_2023_Tesei_IDRome/blob/main/svr_model/svr_model.joblib?raw=true')\n",
    "if os.path.exists('residues.csv') == False:\n",
    "    wget.download('https://raw.githubusercontent.com/KULL-Centre/CALVADOS/main/residues.csv')\n",
    "model = load('svr_model.joblib')\n",
    "residues = pd.read_csv('residues.csv',index_col='one')\n",
    "\n",
    "fasta_dict = {}\n",
    "df = pd.DataFrame(columns=['nu_SVR','fK','fR','fE','fD','fARO','Mean_lambda','SHD','Omega_ARO','SCD','kappa','FCR','NCPR'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "1citqFEJiW-g"
   },
   "outputs": [],
   "source": [
    "#@title <b>Upload sequence(s)</b>\n",
    "#@markdown Upload fasta file. File with multiple sequences and multiple files upload is supported.\n",
    "current_upload = []\n",
    "fasta_file = files.upload()\n",
    "for fn in fasta_file.keys():\n",
    "    fasta = open(fn).readlines()\n",
    "    try:\n",
    "        fasta.remove(\"\")\n",
    "    except:\n",
    "        pass\n",
    "    for l in fasta:\n",
    "        if l.startswith('>'):\n",
    "            name = l[1:].strip()\n",
    "            fasta_dict[name] = ''\n",
    "            current_upload.append(name)\n",
    "        else:\n",
    "            fasta_dict[name] += l.strip()\n",
    "\n",
    "#check sequence\n",
    "for x in current_upload:\n",
    "    for a in fasta_dict[x]:\n",
    "        if a not in aa:\n",
    "            print('WARNING: {} sequence contains a character ({}) not recognized as an aminoacid. This sequence will be ignored.'.format(x,a))\n",
    "            del fasta_dict[x]\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "-I_Hwtk5jm4s"
   },
   "outputs": [],
   "source": [
    "#@title <b>Input sequence(s)</b>\n",
    "#@markdown Or paste a sequence and provide a name. This cell can be executed multiple times to register more sequences.\n",
    "NAME = \"\" #@param {type:\"string\"}\n",
    "SEQUENCE = \"\" #@param {type:\"string\"}\n",
    "\n",
    "if NAME != \"\" and SEQUENCE != \"\":\n",
    "    if \" \" in SEQUENCE:\n",
    "        SEQUENCE = ''.join(SEQUENCE.split())\n",
    "    fasta_dict[NAME] = SEQUENCE\n",
    "\n",
    "    #check sequence\n",
    "    for a in fasta_dict[NAME]:\n",
    "        if a not in aa:\n",
    "            print('WARNING: {} sequence contains a character not recognized as an aminoacid. This sequence will be ignored'.format(name))\n",
    "            del fasta_dict[NAME]\n",
    "\n",
    "else:\n",
    "    print('No NAME and/or SEQUENCE provided. Upload fasta files with the cell above or paste a sequence at the time here.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "jcvD3RXjoKj-"
   },
   "outputs": [],
   "source": [
    "#@title <b>Predict $\\nu$\n",
    "#@markdown Use this cell to calculate sequence features and predict the scaling exponent $\\nu$. Results will be download in a csv file.\n",
    "\n",
    "f = IntProgress(min=0, max=len(fasta_dict), description='Progress:', bar_style='warning')\n",
    "display(f)\n",
    "\n",
    "for k in fasta_dict.keys():\n",
    "    res = calc_seq_prop(fasta_dict[k],residues,1,1,0)\n",
    "    features = np.array([res[8],res[6],res[9],res[10],res[5]])\n",
    "    nu = np.around(model.predict(features.reshape(1, -1)),3)\n",
    "    df.loc[k] = np.concatenate((nu,res))\n",
    "    f.value += 1\n",
    "\n",
    "clear_output()\n",
    "df.to_csv('nupred.csv',index_label='name')\n",
    "files.download('nupred.csv')\n",
    "\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
